ddht/v5_1/alexandria/client.py

```diff
@@ -110,6 +131,93 @@ async def _request(
             )
         return response  # type: ignore
 
+    def subscribe(
+        self,
+        message_type: Type[TAlexandriaMessage],
+        endpoint: Optional[Endpoint] = None,
+        node_id: Optional[NodeID] = None,
+    ) -> AsyncContextManager[
+        trio.abc.ReceiveChannel[InboundMessage[TAlexandriaMessage]]
+    ]:
+        return self.subscription_manager.subscribe(message_type, endpoint, node_id,)  # type: ignore
+
+    @asynccontextmanager
+    async def subscribe_request(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        request: AlexandriaMessage[Any],
+        response_message_type: Type[TAlexandriaMessage],
+        request_id: Optional[bytes],
+    ) -> AsyncIterator[trio.abc.ReceiveChannel[InboundMessage[TAlexandriaMessage]]]:
+        if request_id is None:
+            request_id = self.network.client.request_tracker.get_free_request_id(
+                node_id
+            )
+
+        self.logger.debug(
+            "Sending request: %s with request id %s", request, request_id.hex(),
+        )
+
+        send_channel, receive_channel = trio.open_memory_channel[
+            InboundMessage[TAlexandriaMessage]
+        ](256)
+
+        with self.request_tracker.reserve_request_id(node_id, request_id):
+            async with trio.open_nursery() as nursery:
+                nursery.start_soon(
+                    functools.partial(
+                        self._manage_request_response,
+                        node_id,
+                        endpoint,
+                        request,
+                        response_message_type,
+                        send_channel,
+                        request_id,
+                    )
+                )
+                try:
+                    async with receive_channel:
+                        yield receive_channel
+                finally:
+                    nursery.cancel_scope.cancel()
+
+    async def _manage_request_response(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        request: AlexandriaMessage[Any],
+        response_message_type: Type[AlexandriaMessage[Any]],
+        send_channel: trio.abc.SendChannel[InboundMessage[AlexandriaMessage[Any]]],
+        request_id: bytes,
+    ) -> None:
```

Comment:
No recommendation now, but I bet this method gets simpler if `SubscriptionManager` supported allowed filtering by request_id, or timeouts, or if it accepted the channel it should be sending messages to. I bet that pulling messages from one channel and doing a little work and forwarding them to another will be common enough that spending some time to work out an interface which makes this method shorter is worth it.

Reply:
I'd be happy to clean any of this up.  It's all working pretty well right now but also, it's been developed in a vacuum with me as the only consumer of these APIs (and also having written them), so new :eyes: are appreciated.